tests/lax_scipy_sparse_test.py

```diff
@@ -178,6 +209,201 @@ def tree_unflatten(cls, aux_data, children):
     actual, _ = jax.scipy.sparse.linalg.cg(A, b)
     self.assertAllClose(expected, actual.value)
 
+  # GMRES
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}".format(jtu.format_shape_dtype_string(shape, dtype)),
+       "shape": shape, "dtype": dtype}
+      for shape in [(2, 2)]
+      for dtype in float_types + complex_types))
+  def test_gmres_on_small_fixed_problem(self, shape, dtype):
+    """
+    GMRES gives the right answer for a small fixed system.
+    """
+    A = jnp.array(([[1, 1], [3, -4]]), dtype=dtype)
+    b = jnp.array([3, 2], dtype=dtype)
+    x0 = jnp.ones(2, dtype=dtype)
+    restart = 2
+    maxiter = 1
+
+    @jax.tree_util.Partial
+    def A_mv(x):
+      return matmul_high_precision(A, x)
+    tol = A.size * jnp.finfo(dtype).eps
+    x, _ = jax.scipy.sparse.linalg._gmres(
+        A_mv, b, x0=x0, tol=tol, atol=tol, restart=restart, maxiter=maxiter)
+    solution = jnp.array([2., 1.], dtype=dtype)
+    self.assertAllClose(solution, x)
+
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}_preconditioner={}_qr_mode={}".format(
+            jtu.format_shape_dtype_string(shape, dtype),
+            preconditioner,
+            qr_mode),
+       "shape": shape, "dtype": dtype, "preconditioner": preconditioner,
+       "qr_mode": qr_mode}
+      for shape in [(3, 3)]
+      # TODO(shoyer): get working for np.complex128 and qr_mode=True
+      for dtype in [np.float64]
+      for preconditioner in [None, 'identity', 'exact']
+      for qr_mode in [False]))
+  def test_gmres_against_scipy(self, shape, dtype, preconditioner, qr_mode):
```

Comment:
This new test (verifying consistency with scipy) currently only passes for float64 (not complex128) and only for qr_mode=False. I am concerned that there is bug in the GMRES implementation for these other cases.

We should still merge this PR to facilitate iteration, but my suggestion is that we keep `gmres` out of the main `jax.scipy.sparse` namespace until we fix this.

Reply:
It's not clear to me that the code necessarily should agree with SciPy for the unconverged solution. Why is this?

Reply:
Assuming exact precision arithmetic, shouldn't GMRES be deterministic algorithm on non-pathological inputs?

These tests are running on small matrices with double precision, so loss of numerical precision cannot explain the difference.

Reply:
I'm honestly not totally sure. Different choices of orthogonalization procedure for the Arnoldi process could perhaps influence the basis found. But perhaps you are right.

I suppose an implication of your results is that the qr_mode choices don't agree with one another. So that's something to understand as well.

Reply:
GMRES could still be useful even if restricted to `qr_mode=False` and `dtype=float64`. Perhaps that's a better intermediate state of affairs, but I'm going to try to merge this now so it's easier to iterate.

I suspect there may be two lingering bugs:

- Handling of complex numbers. Perhaps we're missing a complex conjugate somewhere, e.g., in the orthogonalization step?
- Something in QR mode.

Reply:
Indeed, there were a couple of lingering bugs! See https://github.com/google/jax/pull/5034 for the fixes.